File: leetcode150/TwoPointer/167_Two_Sum_II_Input_Array_Is_Sorted.java
/**
 * Problem: 167. Two Sum II - Input Array Is Sorted
 * 
 * Given a 1-indexed array of integers `numbers` that is already sorted in non-decreasing order,
 * find two numbers such that they add up to a specific target number.
 * 
 * Constraints:
 * - The array is sorted.
 * - There is exactly one solution.
 * - You may not use the same element twice.
 * - The solution must use only constant extra space.
 * 
 * Approach:
 * - Use the **two-pointer technique** since the array is sorted.
 * - Initialize two pointers:
 *   - `ptr1` at the beginning (index 0)
 *   - `ptr2` at the end (last index)
 * - Iterate while `ptr1 < ptr2`:
 *   - Compute the sum of `numbers[ptr1]` and `numbers[ptr2]`
 *   - If the sum is greater than `target`, move `ptr2` to the left (decrement `ptr2`)
 *   - If the sum is less than `target`, move `ptr1` to the right (increment `ptr1`)
 *   - If the sum matches `target`, return the 1-based indices `[ptr1 + 1, ptr2 + 1]`
 * 
 * Time Complexity: O(n) -> We traverse the array at most once.
 * Space Complexity: O(1) -> No extra space is used.
 */

class Solution {
    public int[] twoSum(int[] numbers, int target) {
        int ptr1 = 0;  // Left pointer (start of array)
        int ptr2 = numbers.length - 1; // Right pointer (end of array)

        while (ptr1 < ptr2) { // Continue until pointers meet
            int sum = numbers[ptr1] + numbers[ptr2]; // Calculate sum of two numbers
            
            if (sum > target) {
                ptr2--; // Move right pointer left to reduce sum
            } else if (sum < target) {
                ptr1++; // Move left pointer right to increase sum
            } else {
                return new int[] { ptr1 + 1, ptr2 + 1 }; // Return 1-based indices
            }
        }

        return new int[] { -1, -1 }; // Edge case (should never be reached as one solution is guaranteed)
    }
}

File: leetcode150-ts/Array_string/167_Two_Sum_II_Input_Array_Is_Sorted.ts
/**
 * Finds two numbers in a sorted array that add up to a target and returns their 1-based indices.
 * 
 * Approach:
 * 1. Use a two-pointer technique where one pointer (`left`) starts at the beginning (smallest element)
 *    and the other (`right`) starts at the end (largest element) of the sorted array.
 * 2. Calculate the sum of the elements at the two pointers:
 *    - If the sum equals the target, return the 1-based indices of the two elements.
 *    - If the sum is greater than the target, move the `right` pointer left to decrease the sum.
 *    - If the sum is less than the target, move the `left` pointer right to increase the sum.
 * 3. Repeat until the two pointers meet, ensuring we explore all possible pairs efficiently.
 * 
 * Intuition:
 * - The array is sorted, so moving the `left` pointer right increases the sum, while moving the `right`
 *   pointer left decreases the sum. This allows us to efficiently narrow down to the correct pair.
 * - The two-pointer technique exploits the sorted property to avoid unnecessary checks, making it optimal.
 * 
 * Time Complexity (TC): O(n), where n is the number of elements in the array. In the worst case, we traverse the array once.
 * Space Complexity (SC): O(1), as we use a constant amount of extra space (two pointers).
 * 
 * Problem Link: https://leetcode.com/problems/two-sum-ii-input-array-is-sorted/
 * 
 * @param numbers A 1-indexed sorted array of integers in non-decreasing order.
 * @param target The target sum to achieve.
 * @returns An array of two 1-based indices [index1, index2] whose values sum to the target.
 */
function twoSum(numbers: number[], target: number): number[] {
    let left = 0; // Pointer starting at the beginning (0-based index)
    let right = numbers.length - 1; // Pointer starting at the end (0-based index)

    while (left < right) {
        const sum = numbers[left] + numbers[right];
        if (sum > target) {
            right--; // Decrease the sum by moving the right pointer left
        } else if (sum < target) {
            left++; // Increase the sum by moving the left pointer right
        } else {
            return [left + 1, right + 1]; // Return 1-based indices when sum matches target
        }
    }

    return [-1, -1]; // Return [-1, -1] if no solution found (though problem states there is one)
}

// Example Test Cases
console.log(twoSum([2, 7, 11, 15], 9)); // Output: [1, 2] (2 + 7 = 9)
console.log(twoSum([2, 3, 4], 6));      // Output: [1, 3] (2 + 4 = 6)
console.log(twoSum([-1, 0], -1));       // Output: [1, 2] (-1 + 0 = -1)